File: isabelle-ablator/rust/src/uses.rs
//! Single-file usage analysis for `--delete-lemmas`. Mirrors the rocq ablator's
//! `uses.ml`. For each top-level theory goal we record its block span range,
//! whether it closes normally (a real fact, not `oops`), whether its header
//! carries an automation attribute (`lemma foo [simp]:`), the names cited in its
//! proof, and whether the name is ever used *outside* a proof body (which would
//! make deletion unsafe).
//!
//! Correct-by-construction: deletable only when every occurrence of the name
//! outside its own block sits inside an ablatable proof body AND its header has
//! no attribute — so deleting it + whole-proof-ablating its users leaves a
//! well-formed theory. (Standalone `declare foo [simp]` / `lemmas` commands name
//! the lemma → counted as non-proof uses.)

use crate::keyword as kw;
use crate::span::Span;
use crate::token::Kind;

pub struct Lemma {
    pub name: String,
    pub opener: usize,    // span index of the goal statement
    pub block_end: usize, // span index just past the terminator
    pub users: Vec<usize>,
    pub eligible: bool,
}

const MIN_NAME_LEN: usize = 3;

fn names_in(content: &[crate::token::Token]) -> Vec<String> {
    let mut out = Vec::new();
    for t in content {
        if t.is_ident() || t.kind == Kind::LongIdent {
            out.push(t.source.clone());
            if let Some(dot) = t.source.rfind('.') {
                if dot + 1 < t.source.len() {
                    out.push(t.source[dot + 1..].to_string());
                }
            }
        }
    }
    out
}

// `lemma foo [simp]: …` — a bare `[` before the first bare `:` is an attribute.
fn has_attribute(span: &Span) -> bool {
    for t in &span.content {
        if t.source == ":" {
            return false;
        }
        if t.source == "[" {
            return true;
        }
    }
    false
}

struct Goal {
    opener: usize,
    end: usize,
    name: String,
    normal: bool, // closed by a real qed (not oops)
    attr: bool,
    cited: std::collections::HashSet<String>,
}

/// `aggressive` relaxes the normal-close requirement (BE only; caller validates
/// with `--check-build`). The non-proof-use guard is never relaxed.
pub fn analyze(spans: &[Span], aggressive: bool) -> Vec<Lemma> {
    let n = spans.len();
    let mut goals: Vec<Goal> = Vec::new();
    let mut nonproof: Vec<(String, usize)> = Vec::new(); // (name, span index)
    let mut i = 0;
    while i < n {
        let is_goal = spans[i].keyword_kind().map(kw::is_theory_goal).unwrap_or(false);
        if is_goal {
            let opener = i;
            let name = crate::ablate::goal_name(&spans[i]);
            let attr = has_attribute(&spans[i]);
            for nm in names_in(&spans[i].content) {
                nonproof.push((nm, opener)); // the statement is non-proof context
            }
            i += 1;
            let mut depth = 1i32;
            let mut normal = false;
            let mut cited = std::collections::HashSet::new();
            while i < n && depth > 0 {
                let s = &spans[i];
                for nm in names_in(&s.content) {
                    cited.insert(nm);
                }
                match s.keyword_kind() {
                    Some(k) if kw::is_theory(k) => depth = 0, // theory boundary (malformed)
                    Some(k) => {
                        if kw::is_proof_goal(k) || k == kw::PRF_OPEN {
                            depth += 1;
                        } else if kw::is_proof_close(k) {
                            depth -= 1;
                            if depth == 0 {
                                normal = true;
                            }
                        } else if kw::is_qed_global(k) {
                            depth = 0; // oops: no fact added
                        }
                    }
                    None => {}
                }
                i += 1;
            }
            goals.push(Goal { opener, end: i, name, normal, attr, cited });
        } else {
            for nm in names_in(&spans[i].content) {
                nonproof.push((nm, i));
            }
            i += 1;
        }
    }

    goals
        .iter()
        .map(|g| {
            let users: Vec<usize> = goals
                .iter()
                .filter(|g2| g2.opener != g.opener && g2.cited.contains(&g.name))
                .map(|g2| g2.opener)
                .collect();
            let only_in_proofs = nonproof
                .iter()
                .all(|(nm, j)| *nm != g.name || (*j >= g.opener && *j < g.end));
            let eligible = (g.normal || aggressive)
                && !g.attr
                && g.name.chars().count() >= MIN_NAME_LEN
                && !users.is_empty()
                && only_in_proofs;
            Lemma { name: g.name.clone(), opener: g.opener, block_end: g.end, users, eligible }
        })
        .collect()
}

File: rocq-ablator/lib/uses.ml
(* Single-file usage analysis for [--delete-lemmas]. For each top-level goal we
   record the span range of its whole block, whether it is `Qed`-opaque, the
   names cited in its proof body, and — crucially — whether the lemma's name is
   ever used *outside a proof body* (a statement / definition / notation / hint),
   which would make deleting it unsafe.

   Correct-by-construction: a lemma is deletable only when every occurrence of
   its name outside its own block sits inside an ablatable proof body, so
   deleting it + whole-proof-ablating its users leaves a well-formed file. (In
   Coq, an attributed decl like `#[…] Lemma` is not even classified as a goal, so
   automation-registered lemmas are excluded for free; standalone `Hint …`/
   `Existing Instance …` commands name the lemma and count as non-proof uses.) *)

type lemma = {
  name : string;
  opener : int; (* span index of the goal opener (statement) *)
  block_end : int; (* span index just past the terminator *)
  qed : bool; (* terminator is `Qed` (opaque) *)
  users : int list; (* opener indices of in-file proofs citing this lemma *)
  stmt_names : string list; (* names referenced in the statement (for slice closure) *)
  body_names : string list; (* names cited in the proof body (for slice closure) *)
  eligible : bool;
}

let min_name_len = 3

(* every ident in [content], plus the last component of a qualified name *)
let names_in (content : Token.t list) : string list =
  List.concat_map
    (fun (t : Token.t) ->
      if Token.is_ident t then
        match String.rindex_opt t.source '.' with
        | Some i when i + 1 < String.length t.source ->
            [ t.source; String.sub t.source (i + 1) (String.length t.source - i - 1) ]
        | _ -> [ t.source ]
      else [])
    content

type goal = {
  g_opener : int;
  g_end : int;
  g_name : string;
  g_qed : bool;
  g_cited : (string, unit) Hashtbl.t; (* names cited in the proof body *)
}

(* [aggressive] relaxes the `Qed`-opaque requirement (BE only; the caller must
   validate with --check-build). The non-proof-use guard (a) is never relaxed —
   a use in a statement/definition can't be repaired by ablation. *)
let analyze ?(aggressive = false) (spans : Span.t array) : lemma list =
  let n = Array.length spans in
  let goals = ref [] in
  (* non-proof occurrences: (name, span-index) for statements / non-goal spans *)
  let nonproof = ref [] in
  let add_nonproof j =
    List.iter (fun nm -> nonproof := (nm, j) :: !nonproof) (names_in spans.(j).content)
  in
  let i = ref 0 in
  while !i < n do
    if Span.is_goal spans.(!i) then begin
      let g = !i in
      let nm = Span.name spans.(g) in
      add_nonproof g; (* the statement is non-proof context *)
      incr i;
      let depth = ref 1 in
      let cited = Hashtbl.create 16 in
      let qed = ref false in
      let stop = ref false in
      while (not !stop) && !i < n do
        let s = spans.(!i) in
        if Span.is_goal s && !depth = 1 then stop := true (* malformed: leave for outer *)
        else if Span.is_terminator s && !depth = 1 then begin
          qed := Span.head s = "Qed";
          incr i;
          stop := true
        end
        else begin
          List.iter (fun nm -> Hashtbl.replace cited nm ()) (names_in s.content);
          (match s.kind with Span.Open -> incr depth | Span.Close -> decr depth | _ -> ());
          incr i
        end
      done;
      goals := { g_opener = g; g_end = !i; g_name = nm; g_qed = !qed; g_cited = cited } :: !goals
    end
    else begin
      add_nonproof !i;
      incr i
    end
  done;
  let goals = List.rev !goals in
  let nonproof = !nonproof in
  List.map
    (fun gl ->
      let name = gl.g_name in
      let users =
        List.filter_map
          (fun g' -> if g'.g_opener <> gl.g_opener && Hashtbl.mem g'.g_cited name then Some g'.g_opener else None)
          goals
      in
      (* (a): every non-proof occurrence of [name] is within this lemma's block *)
      let only_in_proofs =
        List.for_all
          (fun (nm, j) -> nm <> name || (j >= gl.g_opener && j < gl.g_end))
          nonproof
      in
      let eligible =
        (gl.g_qed || aggressive) && String.length name >= min_name_len && users <> []
        && only_in_proofs
      in
      let stmt_names = names_in spans.(gl.g_opener).content in
      let body_names = Hashtbl.fold (fun k () acc -> k :: acc) gl.g_cited [] in
      { name; opener = gl.g_opener; block_end = gl.g_end; qed = gl.g_qed; users;
        stmt_names; body_names; eligible })
    goals
